### ISSUE TITLE
CPU Power consumption

### ISSUE THREAD

**Issue by velaar** [tylernguyen/x1c6-hackintosh#28]
@tylernguyen 
Can you please confirm the CPU power of about 2W on PKG when idle? If possible also post an Intel power gadget screenshot.

I've got a feeling that the CPU power vectors are not exactly correct. I have a way lower PKG on Linux/Windows (<1W) and trying to pinpoint the issue.

Thanks!

![Screen Shot 2020-01-10 at 8 22 24 AM](https://user-images.githubusercontent.com/4062230/72155986-5864fd80-3382-11ea-9925-5e498c6840bd.png)

**Comment by tylernguyen** [comment]
I just quickly checked, mine is also pulling 2W on idle. Though I'm not exactly sure how to approach this issue. @velaar I saw your pull request, did power consumption improve after installing NVMEFix?

**Comment by velaar** [comment]
@tylernguyen Yes, but not on the CPU package. The system can now go down from around 9W to around 7W and I have a very low power SSD.  

For the CPU package consumption, so far every app I run reports the original 1.3GHz CPU frequency vector. One way would be to determine if that is the case. Also for i7-8650U the minimum frequency is 400MHz but I was unable to patch the vector to push it this low. 

```
CPU Ratio Info:
------------------------------------------
Base Clock Frequency (BLCK)............. : 100 MHz
Maximum Efficiency Ratio/Frequency.......:  4 ( 400 MHz)
Maximum non-Turbo Ratio/Frequency........: 21 (2100 MHz)
Maximum Turbo Ratio/Frequency............: 42 (4200 MHz)

IGPU Info:
------------------------------------------
IGPU Current Frequency...................:    0 MHz
IGPU Minimum Frequency...................:  300 MHz
IGPU Maximum Non-Turbo Frequency.........:  300 MHz
IGPU Maximum Turbo Frequency.............: 1150 MHz
IGPU Maximum limit.......................: No Limit
```

**Comment by Colton-Ko** [comment]
![Screen Shot 2020-01-11 at 09 25 11](https://user-images.githubusercontent.com/49970432/72196606-63885e00-3454-11ea-9e22-fa262813a397.png)
2W on idle as well

**Comment by velaar** [comment]
@Colton-Ko @tylernguyen 
There is a chance that this is just a different type of measurement (comparing to Linux), but I was able to get <3W of _total_ consumption on low brightness. (here: https://www.reddit.com/r/thinkpad/comments/8aqy5g/x1c6_on_linux_idle_29w_real_battery_life_on_light/). I guess 5W should be achievable in MacOS (that should give us about 10h of battery). 

NVMeFix definitely helps. Now the question is if we can really push the CPU to C10 at 400MHz. 
Another question is why CpuFriendFriend (and other similar apps) all show that the vector is 1.3GHz: 

```
Current Frequency Vector:

1 of 2: 0D (1300 MHz)

Enter the new min hex freq (800Mhz would be 08, 1300Mhz would be 0D):
```

**Comment by Colton-Ko** [comment]
@velaar I have only reached the minimum of 800MHz on i7-8550U in macOS. I used https://github.com/stevezhengshiqi/one-key-cpufriend to create a customized DataProvider for CPUFriend.kext

**Comment by velaar** [comment]
I think I did it :) 

![Screen Shot 2020-01-14 at 8 07 40 PM](https://user-images.githubusercontent.com/4062230/72396291-382e9900-370a-11ea-93f2-e9cdcae6dbf9.png)

**Comment by tylernguyen** [comment]
@velaar That's great! What did you do?

**Comment by velaar** [comment]
NVMeFix was the key to this as well as a low power SSD. Plus undervolting (-83mV), using CPU vectors from MacBookPro15,4 adjusted to 400MHz (practically only drops to 600), enabling ASPM through PCIutils on most of the devices, a lot of USB power tweaking.

I've hit a rough patch at work, so can't contribute at a previous rate for some time.

**Comment by velaar** [comment]
Sorry for double-posting. Today I woke my laptop and for some reason it was at an unusual power state. I don't know what caused it (ec bug?) but for some reason the CPU power consumption was exactly what it is supposed to be on this cpu. (See screenshots) I'm now wondering what happened and how can we replicate the condition. The good part is that It is possible to achieve 5.5W draw on X1C6 on MacOS X.


![Screen Shot 2020-01-20 at 5 20 25 AM](https://user-images.githubusercontent.com/4062230/72737568-d2109e80-3b6d-11ea-8cb7-e93b6f53a88d.png)

![Screen Shot 2020-01-20 at 10 25 47 AM](https://user-images.githubusercontent.com/4062230/72738278-3c760e80-3b6f-11ea-84f3-20a4ef103a17.png)

**Comment by velaar** [comment]
❗ HELP NEEDED❗ 

@tylernguyen, @Colton-Ko 
Ok. So I found the way to increase the X1C6 battery life to about 9-10h at realistic brightness with low activity (python coding/web kind of tasks) 

There are 3 components to it: 

**1. Undervolting** 
I use VoltageShift (kext + launchctl) but it doesn't seem to persist after sleep, a better undervaluing solution is needed. It is important that this solution can separately manage the Core, GPU and digital IO voltages and can re-apply it on a 1-3 minute basis (otherwise the EC resets offsets) 

**2. Thunderbolt / USB-C** 
Thunderbolt controller causes a C-storn on our laptops ramping the CPU package to 2.2-2.5W 
If you disable it (or put into BIOS-assisted mode), the C-storm ends and the CPU properly clocks down to where it is supposed to be allowing us to reach a way better power state. While I'm writing this message on battery my laptop consumes about 5W of power while WiFI and Bluetooth are ON. 

I will continue to experiment in this field but here are some findings: 
* USB-C does NOT recognize devices but is powered
* You can power on the TB3 on demand with IOElectrify (kextload) at any point after boot
* SSDT-TB3.aml V1 allows to use thunderbolt if the device was plugged in on boot and live unplug without crash 


Further research is needed into TB3 v3.1 and TBTonPCH to see if we can activate TB on demand and USB-C 

Thunderbolt BIOS assist means (as per Lenovo rep):

>  I confirmed with the experts:
> In BIOS Assist Mode, memory resource arraignment is done by BIOS. 
> In Native mode, memory resource arrangement is done by OS.
>  
> There are two different firmwares - one for each mode - as they aren't compatible between the two modes. LVFS should take care of getting you the right one now as mentioned above.

 

**3. ASPM** 
The best way to set the ASPM mode is through a EFI driver. But I'm not good enough in programming to write one. It can be done with setpci.efi but I don't know how to wrap the setpci into a script that can be called by OpenCore to automate the process


I think that it is worth looking into it. At the moment I'm enjoying a very long battery life (better than some Macs) with some shortcomings (no USB-C at the moment, TB3 requires reboot, have to run a script for max power savings), I hope that with your help we can overcome all of them, but some time commitment is needed.

**Comment by tylernguyen** [comment]
@velaar Thank you so much for contribution!!! It means so much to me that others are helping my project.

I'm investigating the matter on my own as we speak.  Can you point to more references on ASPM and setpci.efi? I want to look deeper into that.  

Also, I would suggest repasting the machine with something like Grizzly thermal to improve undervolting capabilities.  

Moving forward, I will track the energy usage of my EFI release folders and include it in my repo. I also ask that you attach an EFI folder and relevant settings each time energy usage is reported. This will help coordinate settings and experiments.

I'll be back soon with further updates!

**Comment by tylernguyen** [comment]
@velaar Also, what are you using to track energy usage? A combination of iStats menu and Intel Power Gadget? I just want to be on the same page.

**Comment by velaar** [comment]
@tylernguyen 
I track the energy with iStat + intel power gadget, that is correct. 

After running this set of settings for a week I can say that a very low power usage is real.. and it's very easy to get used to. 

Unfortunately I'm quite busy this/next week but will try to upload a quick writeup of what needs to be done to achieve this level of power consumption.

**Comment by Porco-Rosso** [comment]
On a i7-8650U, with Volta set to "power-limit" 1 W and one-key-cpufriend set to "1200/1300mhz balance Power" I get the following:
<img width="1440" alt="Screen Shot 2020-01-26 at 5 08 45 PM" src="https://user-images.githubusercontent.com/7024527/73136460-faf9cd80-405e-11ea-9214-8c080b4b2769.png">

Is it safe for me to try 400mhz on cpufriend, or is there a chance I won't be able to boot back into macOS?

**Comment by velaar** [comment]
@Porco-Rosso 
Yes, it is safe, but no, it is verified to not work (no difference from 800MHz). I don't use Volta as it doesn't work that well on ThinkPads (EC constantly resets the voltage offsets, visible if you read MSR directly), maybe they fixed it by now. Will retest soon. 

 Also setting PL1 or PL2 (not sure what power-limit actually means) will kill the CPU performance (lock your CPU down to 400MHz)

**Comment by tylernguyen** [comment]
@velaar Is there any updates on your instructions to maximize batt life? Sorry to bother.

**Comment by AutinMitra** [comment]
When I last checked, my wattage was ~7-9W browsing Chrome and using VSCode for web development, and that's without undervolting.

To highlight ways we can improve PM, we can maybe add a section to README.md?

**Comment by tylernguyen** [comment]
@velaar Is there any updates on your instructions to maximize battery life?

**Comment by lubixxx96** [comment]
Hello.

Is that only me or VoltageShift does not work on our machine?
I get instant crash on Windows when I am trying to apply offset of -100mV or more, but in OSX I can set -200 or more and nothing. I dont even see any temperature differences.
Powerlimits are working fine.

I am using VoltageShift Undervoltage Tool v 1.21. with deamons for reapply every minute.
Any ideas?

Thanks,
Lubomir

**Comment by tylernguyen** [comment]
@lubixxx96 I haven't really tried VoltageShift myself. It was really only here b/c of reference from @velaar . In the next week, I will do some testing on it and get back to you.

**Comment by velaar** [comment]
@lubixxx96 
It does work, however you should be aware of the EC behaviour where it resets TDP to 15W/25W as soon as you hit either ~90C or ~40W in power. That's why reapply is there. 

I'm not sure what "VoltageShift Undervoltage Tool v 1.21" is. I'm using kexts and software from here: https://github.com/sicreative/VoltageShift 

And I crash at about -90mV on the CPU or at about -120mV on the digital/ring side. In Windows I crash at -80mV in linux the best sustained result was -90mV (can be a calculation/rounding error)

@tylernguyen 
Sorry for lack of response here. I'm also a student going through my finals at the moment, should be over by Apr. 16. 

The main points about power consumption that I hope you can look into: 

**Powering up/powering down USB-C/Thunderbolt controller causes an interrupt storm.** It might be on the CPU side or the USB controller side but the power consumption is higher by about 2W on average. There is some discussion of it here https://forums.lenovo.com/t5/Linux-Discussion/FYI-Linux-May-Not-Support-Thunderbolt-Native-Mode/td-p/4057604?page=1 

**USB power management** -  Believe can be tuned further but I don't know how exactly to do that. It seems Mac OS does not completely disable the USB ports when nothing is plugged in 

**APSM** I'm enabling ASPM on most of the devices with a help of a script. I know that is can be done on either UEFI level or with a UEFI/Opencore driver (tested by loading an efi shell and running the  UEFI versions of setpci/lspci to set the registers as follows: 
```
./setpci -s 03:00.0 50.b=3:3
./setpci -s 02:00.0 50.b=3:3
./setpci -s 04:00.0 50.b=3:3 
```


** Thunderbolt **- I don't really expect power management to work there without Apple firmware on the controller. However if we can make a more reliable thunderbolt implementation with "Thunderbolt BIOS assist" ON we might get hotplug again as BIOS does the d device init. We just need to make it look like an External PCIe device. 

**Undervolting can be done on BIOS/UEFI/EC side to achieve permanent results** - that is pending Skyra1n 

[aspm2.sh.zip](https://github.com/tylernguyen/x1c6-hackintosh/files/4475701/aspm2.sh.zip)

**Comment by lubixxx96** [comment]
@velaar I am using same VoltageShift v1.21 kext and app from same github repo. No matter what I set for cpu, it does not crash. Only the cache offset crash, not sure about other (system agency, analogy i/o) OC mailbox returns failed and looks like I cant change Digital I/O - still 0mv.

Thanks for the aspm2 script, I am using it as lauchdaemon.

BTW, do we have an option to manually control fan rpm?

**Comment by tylernguyen** [comment]
@lubixxx96 I do not think so, not until a BIOS release exploit is out.
@velaar Where you getting setpci and lspci binaries from? I cannot find a source to download them.

**Comment by lubixxx96** [comment]
Here you have:
[setpci_lspci.zip](https://github.com/tylernguyen/x1c6-hackintosh/files/4487649/setpci_lspci.zip)

**Comment by velaar** [comment]
@tylernguyen 
I've build pciutils myself. Booth for MacOS and efi. 

Here is what I used to get the setpci.efi https://github.com/timotheuslin/pciutils.efi

**Comment by tylernguyen** [comment]
@velaar Thanks for your binaries. I do have a question though,

Which pic devices specifically are you enabling ASPM for?
And which ASPM mode are you enabling? I assume L1?

**Comment by benbender** [comment]
Just as a quick-note: I'm working on this issue on a T480, which is very similar. Three big parts of the puzzle of power consumption are:

a) Power management of the iGPU. All your screenshots showing a max frequency of ~350mhz which means that the GuC-Firmware, and therefore power management, is missing.

b) USB-Power for internal devices (f.e. Card-Readers) which may be powered up constantly

c) MAYBE: dGPU - Whatevergreen just powers down the PCIe-port and I've disabled Optimus via ACPI instead and power consumptions seems to be improved a bit.

**Comment by tylernguyen** [comment]
> 
> 
> Just as a quick-note: I'm working on this issue on a T480, which is very similar. Three big parts of the puzzle of power consumption are:
> 
> a) Power management of the iGPU. All your screenshots showing a max frequency of ~350mhz which means that the GuC-Firmware, and therefore power management, is missing.
> 
> b) USB-Power for internal devices (f.e. Card-Readers) which may be powered up constantly
> 
> c) MAYBE: dGPU - Whatevergreen just powers down the PCIe-port and I've disabled Optimus via ACPI instead and power consumptions seems to be improved a bit.

@benbender Thanks for letting me know.

The x1c6 does not have a dGPU so that's not an issue.
On the next repo update, I will force load Apple GuC-Firmware which should improve power consumption somewhat. However, I have been experimenting it on my machine and it really hasn't saved that much power.
I suspect that the biggest power hog is USB and Thunderbolt. But I really don't know how to begin to investigate.

**Comment by benbender** [comment]
For USB:
My "attack vector" is patching the ports directly via ACPI: https://github.com/daliansky/OC-little/tree/master/16-ACPI%E5%AE%9A%E5%88%B6USB%E7%AB%AF%E5%8F%A3
For things like IR-Camera etc
But f.e. I would like to use the card reader, but it seems to constantly consume 500mW of USB-Power - As there seems no known way of patch for internal USB-Card-Readers in Catalina these kinds of devices are a real PITA for power consumption.

For Thunderbolt:
My target is to disable TB and retain USB-C 3.1. I've been experimenting with disabling the PCI-E-Interface for TB, but was only able to get USB-C 3.0 @ 5GBit. Time will tell ;)

For GuC:
Is GuC-Loading working for you? 

For me it fails to get the firmware hash from the ME-controller. Therefor igfxfw=2 and loading of linux-guc fails. Additionally there are closed bug-reports on the acidanthera bug-tracker which suggest, that there is only skylake-guc for kabylake in the apple-framebuffers which gives me a KB with igfxfw=1. I wasn't successful on that part in any way. PCIID for my UHD620 is 5917.

**Comment by tylernguyen** [comment]
@benbender Thank you, I'll look into those ACPI patches.
For GuC, igfxfw=2 works fine for me, and I have the same iGPU as you. What's your framebuffer settings?

**Comment by benbender** [comment]
I'm using 591C0005 with the following config:
```
                        <key>PciRoot(0x0)/Pci(0x2,0x0)</key>
                        <dict>
                                <key>AAPL,GfxYTile</key>
                                <data>AQAAAA==</data>
                                <key>AAPL,ig-platform-id</key>
                                <data>BQAcWQ==</data>
                                <key>AAPL,slot-name</key>
                                <string>Internal@0,2,0</string>
                                <key>device-id</key>
                                <data>FlkAAA==</data>
                                <key>framebuffer-con2-enable</key>
                                <data>AQAAAA==</data>
                                <key>framebuffer-con2-type</key>
                                <data>AAgAAA==</data>
                                <key>framebuffer-patch-enable</key>
                                <data>AQAAAA==</data>
                                <key>hda-gfx</key>
                                <string>onboard-1</string>
                                <key>model</key>
                                <string>Intel(R) UHD Graphics 620</string>
                        </dict>
```

Accompanied by a binary patch to get rid of the 32mb-dvmt-check, this gives me perfectly stable 4k/60Hz output via DP/USB-C and 30Hz via HDMI 1.4.

I'll try to check out your config and have a look, if that gives me GuC-acceleration. But I fear my problem is related to the (T480-)Firmware and the hash-generation which is used in GuC-Firmware-Loading.

Btw, atm I'm running -disablegfxfirmware which gives me at least 500mhz "boost"-clock on the GPU and doesn't seem to have a negative impact on power consumption.

**Comment by parndt** [comment]
> Accompanied by a binary patch to get rid of the 32mb-dvmt-check, this gives me perfectly stable 4k/60Hz output via DP/USB-C and 30Hz via HDMI 1.4.

@benbender 😲 what is the binary patch you're using? I tried to get 4K@60Hz on USB-C/DP and could never get it working (see #40 if you're interested).

**Comment by benbender** [comment]
````
		<key>Patch</key>
		<array>
			<dict>
				<key>Base</key>
				<string></string>
				<key>Comment</key>
				<string>KBL-DVMTpatch-10.15-10.15.1-10.15.2-10.15.3</string>
				<key>Count</key>
				<integer>0</integer>
				<key>Enabled</key>
				<true/>
				<key>Find</key>
				<data>D4KnAwAA</data>
				<key>Identifier</key>
				<string>com.apple.driver.AppleIntelKBLGraphicsFramebuffer</string>
				<key>Limit</key>
				<integer>0</integer>
				<key>Mask</key>
				<data></data>
				<key>MaxKernel</key>
				<string>19.3.0</string>
				<key>MinKernel</key>
				<string></string>
				<key>Replace</key>
				<data>D4IAAAAA</data>
				<key>ReplaceMask</key>
				<data></data>
				<key>Skip</key>
				<integer>0</integer>
			</dict>
			<dict>
				<key>Base</key>
				<string></string>
				<key>Comment</key>
				<string>KBL-DVMTpatch-10.15.4-10.15.5</string>
				<key>Count</key>
				<integer>0</integer>
				<key>Enabled</key>
				<true/>
				<key>Find</key>
				<data>D4KPAwAA</data>
				<key>Identifier</key>
				<string>com.apple.driver.AppleIntelKBLGraphicsFramebuffer</string>
				<key>Limit</key>
				<integer>0</integer>
				<key>Mask</key>
				<data></data>
				<key>MaxKernel</key>
				<string></string>
				<key>MinKernel</key>
				<string>19.4.0</string>
				<key>Replace</key>
				<data>D4IAAAAA</data>
				<key>ReplaceMask</key>
				<data></data>
				<key>Skip</key>
				<integer>0</integer>
			</dict>
			<dict>
				<key>Base</key>
				<string></string>
				<key>Comment</key>
				<string>CFL-DVMTpatch-10.15-10.15.1-10.15.2-10.15.3</string>
				<key>Count</key>
				<integer>0</integer>
				<key>Enabled</key>
				<true/>
				<key>Find</key>
				<data>D4KnAwAA</data>
				<key>Identifier</key>
				<string>com.apple.driver.AppleIntelCFLGraphicsFramebuffer</string>
				<key>Limit</key>
				<integer>0</integer>
				<key>Mask</key>
				<data></data>
				<key>MaxKernel</key>
				<string>19.3.0</string>
				<key>MinKernel</key>
				<string></string>
				<key>Replace</key>
				<data>D4IAAAAA</data>
				<key>ReplaceMask</key>
				<data></data>
				<key>Skip</key>
				<integer>0</integer>
			</dict>
			<dict>
				<key>Base</key>
				<string></string>
				<key>Comment</key>
				<string>CFL-DVMTpatch-10.15.4-10.15.5</string>
				<key>Count</key>
				<integer>0</integer>
				<key>Enabled</key>
				<true/>
				<key>Find</key>
				<data>D4KPAwAA</data>
				<key>Identifier</key>
				<string>com.apple.driver.AppleIntelCFLGraphicsFramebuffer</string>
				<key>Limit</key>
				<integer>0</integer>
				<key>Mask</key>
				<data></data>
				<key>MaxKernel</key>
				<string></string>
				<key>MinKernel</key>
				<string>19.4.0</string>
				<key>Replace</key>
				<data>D4IAAAAA</data>
				<key>ReplaceMask</key>
				<data></data>
				<key>Skip</key>
				<integer>0</integer>
			</dict>
		</array>
`````

These are working fine for bothe, the KBL/CFL-Framebuffers. But you need to have 32mb-stolenmem-patches in place for updates and recovery. Booting in update-/recovery-mode OC isn't able to apply those patches and will result in a kernel-panic otherwise. Additionally you may need to update kextcaches ("sudo kextcache -i /") if you change ig-platform-id. If caches and config are out of sync, these may also lead to KP. (At least as far as I can tell). Oh, and "AvoidFastLinkTraining" should be disabled in your framebuffer of choice. Otherwise 4k was a bit unstable for me from time to time.

But as my daily driver with a 4k-32"-Display it's working perfectly stable, fine and even fast.

**Comment by parndt** [comment]
Thanks @benbender - I'll give that a go when I get a chance.

**Comment by benbender** [comment]
You're welcome! If needed, I could share my config.plist as a whole. I don't think that there are more parts involved (besides WEG-config and the patches - and maybe the platform-id of macbookpro15,2), but I'm not entirely sure.

**Comment by benbender** [comment]
@tylernguyen just tried with macbook14,1, exactly the framebuffer-config from your repo and igfxfw=2 and got stuck again in a failed loop of

````
[IGPU] Hash data from ME never returned, status = 1, doing retry #
````

So I think my guess is right and it's firmware/bios related. You may want to put a note in the docs if you plan to update your repo :/

**Comment by tylernguyen** [comment]
> You're welcome! If needed, I could share my config.plist as a whole. I don't think that there are more parts involved (besides WEG-config and the patches - and maybe the platform-id of macbookpro15,2), but I'm not entirely sure.

Actually, could you share your EFI zip as a whole? I want to cross reference everything possible. Make sure to remove your SMBIOS info before sharing.

**Comment by velaar** [comment]
@tylernguyen can you share your latest config.plist too? I struggle to enable igfwfx. Fails to load the WEG altogether

**Comment by benbender** [comment]
Sure - It's not cleaned up and there are some experiments in there, but the config.plist is mostly stable for me.

Here you go: [OC-T480.zip](https://github.com/tylernguyen/x1c6-hackintosh/files/4828288/OC-T480.zip)

**Comment by tylernguyen** [comment]
@benbender  This is embarrassing. I misspelled `igfxfw` as `ifgwfx`. With the correct spelling, it fails to load. Thanks for double-checking me. @velaar

**Comment by benbender** [comment]
My testing was done with the following bootargs:
````
 <string>debug=0x100 keepsyms=1 brcmfx-country=#a brcmfx-driver=2 -wegnoegpu -wegdbg -v igfxfw=2 </string>
````

So it may be an embarrassing (wouldn't say that ;) mistake, but not my cause of the problem. No need to worry!

**Comment by tylernguyen** [comment]
> 
> 
> My testing was done with the following bootargs:
> 
> ```
>  <string>debug=0x100 keepsyms=1 brcmfx-country=#a brcmfx-driver=2 -wegnoegpu -wegdbg -v igfxfw=2 </string>
> ```
> 
> So it may be an embarrassing (wouldn't say that ;) mistake, but not my cause of the problem. No need to worry!

So if we can't load Apple GuC firmware. How has your experience with `-disablegfxfirmware` been?
Also, with your kernel patches? Do you just disable them every time you need to enter recovery or do macOS updates?

**Comment by benbender** [comment]
Regarding iGPU-FW:
No boot-args regarding firmware (which results in igfxfw=4 afaik) gives me a maximum of 300mhz and no power-management on the iGPU.

-disablegfxfw gives me a maximum of 500mhz and no power-management on the iGPU.

-igfxfw=1 & -igfxfw=2 leads to a loop of failed hash-retrieval and finally no fw-load.

For the patches:
No need to disable them at any point. They are simply not applied for recovery/updates - so you need to have the framebuffer-stolensize patches to not crash into a KP.

And there is the problem, in case of updates, that the framebuffers could change and the patches don't fit anymore (which happened with 10.15.14 as you see above).

My strategy on this is to test new OSX point-releases with my config on an external ssd-enclosure with a clean install and my config beforehand (not only because of the patches, but because the t480 is my daily driver) and apply them if it seems to be working stable

**Comment by tylernguyen** [comment]
@benbender Also to answer your previous question, the reason I'm using `MacbookPro14,1` is mostly because this repo/project was started over 2 years ago. And if I can recall, `MacbookPro15,2` was not an available or stable SMBIOS at the time. `MacbookPro14,1` has been perfectly stable for me since so I never bother to experiment with other SMBIOS. So I guess I should start trying some different ones out soon. Have you ever tried `MacbookPro14,1` and in your experience, does the SMBIOS affect CPU Power Consumption, with CPUFriendDataProvider applied?

**Comment by tylernguyen** [comment]
Edit: Also, `MacbookPro15,2` is a Coffee Lake processor while `MacbookPro14,1` is a Kaby Lake one which matches closest to my machine. I'll experiment with different SMBIOS soon but I do not think using one of a different platform could possibly be better?

**Comment by benbender** [comment]
Not quite sure - I've done to much experimentation in this regard in parallel.

What's an additional PITA is the fact, that sensor readings not the same between platforms. f.e. iStat Menus maps the sensor-values to static names of according to platform (for macbookpro15,2 the cpu-sensors start 0-indexed f.e. and my cpu-core 4 is shown as "CPU Proximity". Same for power readings).

**Comment by tylernguyen** [comment]
@benbender 
What's your CPU? is it the Coffee Lake lake UHD 620 or the Kaby Lake -R UHD 620?

**Comment by benbender** [comment]
@tylernguyen thats not quite accurate. Kaby lake is essentially 7xxx-series in mac os-world, coffee lake is 8xxx-series and kaby lake refesh (which is hat the 8650u is) isn't available in the mac-world. There are a number of people who are even forcing lilu to lilucpu=9 for kbr.

My CPU is kaby lake refresh 8650u with an (U)HD 620 8086-5917. 

But these namings are generally more marketing than anything else and, at times, generations of cpus/gpus even get mixed when it fits budgets/marketing/etc. 

See f.e.: https://github.com/acidanthera/bugtracker/issues/290

**Comment by benbender** [comment]
Oh and, as I said, I couldn't really find any meaningful explanation of the effects of platform-id choice. Some say it should be chosen by gpu, some cpu.

In my opinion it's more about cpu, because frequency-vecors/cpu-pm really depends on that choice (but can be altered). Besides of that, different assumptions of os x and/or apps seem to depend on it (f.e. the expected sensors of iStat, TouchID etc)

**Comment by tylernguyen** [comment]
@benbender Right, but the problem is that a lot of our `config.plist` variables depend on Kaby Lake properties, not Coffee Lake ones. For examples, your `device-id` is 1659000 (Intel HD Graphics 620) while your `AAPL,ig-platform-id` is `05001C59` (Intel HD Graphics 630), both of which are Kaby Lake graphics. So when it comes to switching a different SMBIOS, it's a lot more work than just changing `MacbookPro14,1` to  `MacbookPro15,2`. For it be to truly optimized and improve anything, you would need to patch various variables to be in sync with your SMBIOS. Or at least that's what I'm assuming? it not even affect anything if the SMBIOS definition and various config variables do not match

**Comment by benbender** [comment]
That's a point of question for me where I'm uncertain. ATM I tend to believe that platform-id/framebuffer do not depend on one another (which is backed by the fact f.e. that it was advised to fake kaby-lake down to skylake when there was no fitting framebuffer in os x).

Besides of that I'm a bit unsure what kind of variables you are talking about (the only ones I'm aware of are those in the platform-id-section of OC and sure, they have to match)

**Comment by benbender** [comment]
fun fact btw: original macbooks have poor thermals and cpu-performance on windows because every optimization is done via kext in osx (freq-vectors, power-profiles, therma-pressure etc) and not available to windows even if it's installed via bootcamp... So, in the end, its all software-based tuning which means a) it can be altered and b) (and thats a kind of a problem for us) the data in the acpi tables is meaningless nowadays to os x/hackintoshing.

**Comment by tylernguyen** [comment]
@benbender 
If you patch framebuffer-conX using `alldata` regressions, hardware flag bytes is matched to the ioreg dump of which platform and their respective graphics. That's a big problem in getting HDMI to work. The worse headache is the thunderbolt problems, each different platform has it own USB specifications due to the number of Thunderbolt ports it has, which is more headache when it comes to patching displayport and USB for thunderbolt.

**Comment by benbender** [comment]
Hm, to be honest, I don't think so.

Alldata isn't matched to anything, but a concatenated string of all the connector variables you could define independently. It doesn't make any difference other than its more error-prone and less readable (at least for me).

Its (if I remember correctly) simply: Index-BusId-Pipe-ConnectorType-FramebufferFlags concatenated.

USB/TB-Ports rely on the names originating from acpi. So I don't know about any relevance of the platform-id.

**Comment by benbender** [comment]
Oh, I might have misunderstood: if you don't define every variable its "matched" against the defaults of the framebuffer, yes. 

But those are, at least to my knowledge not related to the platform-id in hackintosh-land. You could, technically, have a haswell-framebuffer-kext loaded on icy lake.

**Comment by tylernguyen** [comment]
> 
> 
> Oh, I might have misunderstood: if you don't define every variable its "matched" against the defaults of the framebuffer, yes.
> 
> But those are, at least to my knowledge not related to the platform-id in hackintosh-land. You could, technically, have a haswell-framebuffer-kext loaded on icy lake.

I think you are right. Though the thought of doing that makes me uncomfortable somehow. But again, I think you're correct that it should not matter. So I guess after all, I should change my SMBIOS and test performance/power between the two. 

Though I do not know how to do this objectively. Do you have any ideas on how to compare the two SMBIOS?

**Comment by benbender** [comment]
Not really. If you are willing to invest the time, you should try to find a comparable scenario which is repeatable.

f.e. boot, wait 2 min for the os/cpu to settle and run geekbench, compare max power-usage.

And that should be done with one change at a time. But even than there are more factors like temperature which will interfere. If the difference isn't directly noteable, it will be hard. And I don't think there are big, easy gains - if there were, they would be known.

On the other hand, most of the hackintosh-folks just repeat what is there in terms of knowledge/experience - and these are often outdated, undercomplex or simply false.

F.e. I weg with -wegnoegpu simply shuts down the pcie-port of the egpu. I've had the feeling that it makes a difference to shut down optimus properly. But I didn't do any real meassurements to back up my feeling ;)

**Comment by benbender** [comment]
btw as an example of "outdated" knowledge: for me, sleep/hibernate is working perfectly fine with only one line of acpi-patch, darkwake=4 and womp and proximity-alert disabled (which may activate bt/glan in sleep and not supported anyways). besides of these things, my pmset-configs are untouched and fine (for me), even with darkwake enabled. 

no name6d/gprw-patches or such needed. Those are, if needed at all, highly hardware specific patches.

But you have to acknowledge that some functions of osx are firmware-dependent and simply not available to us (f.e. wakeup by BT-mouse, which would us force to power up the usb-bt-adapter, which would lead to battery drain in sleep... etc)

**Comment by tylernguyen** [comment]
> 
> 
> btw as an example of "outdated" knowledge: for me, sleep/hibernate is working perfectly fine with only one line of acpi-patch, darkwake=4 and womp and proximity-alert disabled (which may activate bt/glan in sleep and not supported anyways). besides of these things, my pmset-configs are untouched and fine (for me), even with darkwake enabled.
> 
> no name6d/gprw-patches or such needed. Those are, if needed at all, highly hardware specific patches.
> 
> But you have to acknowledge that some functions of osx are firmware-dependent and simply not available to us (f.e. wakeup by BT-mouse, which would us force to power up the usb-bt-adapter, which would lead to battery drain in sleep... etc)

I've been experimenting with hibernation but I'm always hit with a CMOS error when I try to hibernate on mode 25. Were you able to get true hibernation to work?

**Comment by benbender** [comment]
I didn't try mode 25. Default on modern macs is 3 (mixed ram-sleep/hibernation to disk) and working perfectly fine for me. I can sleep my notebook over night without any battery drain and with attached usb-devices (which suck battery on an always-on-usb-port... ^^). That were my testcases.

Though I didn't test it for serveral days in sleep in case of drain. But that scenario doesn't matter to me anyway - I do need my notebook too often ;)

**Comment by tylernguyen** [comment]
> 
> 
> I didn't try mode 25. Default on modern macs is 3 (mixed ram-sleep/hibernation to disk) and working perfectly fine for me. I can sleep my notebook over night without any battery drain and with attached usb-devices (which suck battery on an always-on-usb-port... ^^). That were my testcases.
> 
> Though I didn't test it for serveral days in sleep in case of drain. But that scenario doesn't matter to me anyway - I do need my notebook too often ;)

Yeah, mode 3 has been working perfect for me as well. Though mode 25 is what I'm really after. And as far as I know, most hackintosh projects do not have trouble with mode 3 after some small macOS tweaks.

**Comment by benbender** [comment]
May I ask why? Is there any benefit besides of "I've done it!"

**Comment by tylernguyen** [comment]
Mode 3 seems to be draining a decent amount of power for me, more than it does compared to S3 sleep on Windows or Linux anyway. So I wanted to get mode 25 working to see how much power it drains.

**Comment by benbender** [comment]
I can sleep mine in mode 3 for 8 hours, "pmset -g log" is nearly empty between sleep and wake and no visible drain at all.

For reference:
```
❯ pmset -g
System-wide power settings:
 DestroyFVKeyOnStandby		1
Currently in use:
 standbydelaylow      10800
 standby              1
 womp                 0
 halfdim              1
 hibernatefile        /var/vm/sleepimage
 proximitywake        0
 powernap             1
 gpuswitch            0
 networkoversleep     0
 disksleep            10
 standbydelayhigh     86400
 sleep                1
 hibernatemode        3
 ttyskeepawake        1
 displaysleep         10
 tcpkeepalive         1
 highstandbythreshold 50
 lidwake              1
````

And with this, I have to leave ;)

**Comment by benbender** [comment]
On my way to bed: another difference might be that I didn't "disable" my sleepimage. So your system may stay in hibernate-to-ram and may not be able to go to hibernate-to-disk. That could mean that your ram is powered on the whole time and would, at least, attribute to the battery drain.

**Comment by benbender** [comment]
And a last last one: 

a) here we might have another difference between platform-ids - you could compare your pm-caps in 14,1 with mine:

````
❯  pmset -g cap
Capabilities for AC Power:
 displaysleep
 disksleep
 sleep
 womp
 lidwake
 halfdim
 gpuswitch
 standby
 standbydelayhigh
 standbydelaylow
 highstandbythreshold
 powernap
 ttyskeepawake
 hibernatemode
 hibernatefile
 tcpkeepalive
 proximitywake
````

And there seem to be some reports of problems with hibernatemode 25 on genuine macs also. So there might be general bugs beside of the beloved hackintosh.

**Comment by velaar** [comment]
> @benbender Right, but the problem is that a lot of our `config.plist` variables depend on Kaby Lake properties, not Coffee Lake ones. For examples, your `device-id` is 1659000 (Intel HD Graphics 620) while your `AAPL,ig-platform-id` is `05001C59` (Intel HD Graphics 630), both of which are Kaby Lake graphics. So when it comes to switching a different SMBIOS, it's a lot more work than just changing `MacbookPro14,1` to `MacbookPro15,2`. For it be to truly optimized and improve anything, you would need to patch various variables to be in sync with your SMBIOS. Or at least that's what I'm assuming? it not even affect anything if the SMBIOS definition and various config variables do not match

This chain got a little bit lost with hibernation-related messages. The question for this issue at the moment is should we move the repo to 15,2 SMBIOS? If yes, is it worth the hassle with iMessage/FT/AAPL integration.

The power consumption is still an issue and the best of my knowledge it is not as much about CPU/Graphics but Thunderbolt/USB3.2 power management (that causes the C-states to go up)

**Comment by tylernguyen** [comment]
@velaar @benbender 
There also exists more options: `MacbookAir8,1` with its graphics injection of the UHD617 may be kinder on CPU power consumption as well. Also remember that changing SMBIOS will also mean that we need to generate a new USBPort patch each platform (which should be easy).

Again, I want to test this in a fair way. But I don't really know how the best method to compare SMBIOS performance and power consumption.

**Comment by benbender** [comment]
Hey,

@velaar For me, CPU-performance and power-consumption is quite good with 15,2 without any real hassle. It's not that much of a pain to change, but I tend to think It doesn't matter that much.

Could you give me a pointer how to diagnose the power-management/c-states problem?
I know that there was a bug in the TB-FW on many Thinkpads in 2018, but it's fixed afaik.


@tylernguyen I didn't consider MacbookAir8,1 at all before, but I don't think, I will investigate in that direction any time soon.

I tested a configuration a few days back where I pushed CPU/GPU to be coffee-lake (lilucpu=10, whatevergreen). It did run without any big trouble (besides different PNLF-Registers, so backlight-controls didn't work without patching), but I didn't saw much of a difference. So I'm tending to believe that that the plattform-id might be a minor factor of PM.

As I said, I will most likely give up testing in that direction and concentrate on PCI-E/USB as my CPU-Stats are looking OK to me (but not "perfect"). I think that usb-suspense of the webcam/cardreader are a real-problem and the thunderbolt-controller is still an issue too.

BTW, I had a look on an macbook-pro15,1 ioreg and ASPM was only activated on WIFI and NVME. So I assume, that's OK.

my basic problem with power management atm is, that, from my knowledge, there is no tool like powertop on linux to really meassure power consumption of the hardware (mainly pcie/usb-devices) on OSX. Without that, we are kinda blind on testing. In my opinion the focus on CPU-Watts might be a bit false leading, when things like the cardreader are eating 500mW constantly and we don't have a good way of debug those problems.

**Comment by 07151129** [comment]
> my basic problem with power management atm is, that, from my knowledge, there is no tool like powertop on linux to really meassure power consumption of the hardware (mainly pcie/usb-devices) on OSX. 

It does have `powermetrics`.

**Comment by benbender** [comment]
@07151129 yes, it does. But f.e. I don't know of any way to debug the real power consumption of an internal USB-device (like an USB-cardreader). It's shows mainly Interrupts and C-States afaik - Or am I missing somethong?

**Comment by 07151129** [comment]
It is capable of computing device power state residency, which closely correlates with the energy consumption by a device. In theory, those numbers can be translated to Joules, but very few drivers publish power consumption metrics. It still useful if you want to ensure that your device goes to low-power states when expected.

**Comment by benbender** [comment]
That sounds interesting. May I ask if you have any links to read further? I didn't found any possibility to debug power-usage of f.e. USB-Devices, just processes, GPU/CPU and ACPI-Devices.

My actual thought was like this: we can debug power-usage of the CPU. Its between 1-3W for the 8650u in idle. The machine is using 10-15W with low load. There are many other power-drawing sources in the machine for which we don't know of any possibility to debug on OSX.
That's the reasoning for my statement about the missing debug-capabilities. Especially on the X1C6/T480 there might be a constant power-draw by the TB-Controller and the Webcam/Card-Reader which I suspect to be part of the reason for the worse battery-performance compared to Windows - which I'd like to tackle.

Btw: Thanks for your work and help! Much appreciated!

**Comment by 07151129** [comment]
Have you tried `powermetrics --samplers devices`? It shows state residency per device class. This should be traceable back to the actual device IOReg path, but looks like the program won't do that.

**Comment by benbender** [comment]
Yes, but didn't seem to be helpful to me:

````
❯ sudo powermetrics --samplers devices
Password:
Machine model: MacBookPro15,2
SMC version: Unknown
EFI version: 1517.0.0.115.1
OS version: 19F101
Boot arguments: debug=0x100 keepsyms=1 -disablegfxfirmware
Boot time: Tue Jun 30 17:47:43 2020



*** Sampled system activity (Tue Jun 30 18:18:46 2020 +0200) (5002.11ms elapsed) ***


**** Device Power States ****
Device                              State Residencies ms/s
IOPMrootDomain                         0.00     0.00     0.00     0.00  1000.00O
ApplePS2Controller                     0.00     0.00U 1000.00U
AppleACPIPCI                           0.00     0.00  1000.00O    0.00
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleBacklightDisplay                  0.00     0.00     0.00U 1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
AppleIntelFramebuffer                  0.00     0.00  1000.00U
IOUSBMassStorageInterfaceNub           0.00  1000.00U
BroadcomBluetoothHostControllerUSBT 1000.00     0.00L    0.00U
...
````

**Comment by tylernguyen** [comment]
Btw, an interesting project to keep and eye on:
https://github.com/syscl/CPUTune

**Comment by benbender** [comment]
Another one (based on VoltageShift): https://github.com/tctien342/smart-cpu

**Comment by FlasHRender** [comment]
So I have had great CPU power usage and total discharge on i7-8550U in Asus Zenbook S. With thunderbolt(turned on) its 1.7W and without thunderbolt(post sleep - sleep breaks thunderbolt) around 1W(goes lower(0.65W) with wifi and bt turned off).

Attached are with thunderbolt (broken post sleep).  
![coconut battery](https://user-images.githubusercontent.com/5005743/86392763-9e5ba800-bcb9-11ea-9d5b-9dccb1c8dc0b.png)
![intel power](https://user-images.githubusercontent.com/5005743/86392771-a0be0200-bcb9-11ea-9777-0102146975e6.png)

**Comment by benbender** [comment]
The 4 Watt overall usage are impressive - mine is around 6 or more :)

May I ask some questions:
- Why do you do combine your dsdt-hotpatches?
- Did you notice any difference with macbookpro16,3-plattform?
- How did you generate your CPUFriendProvider.kext? (afaik most tools doesn't support macbookpro16,x yet)
- What are the AppleRTC-patches for?

**Comment by FlasHRender** [comment]
1) I used to use separate Hotpatches but I could never get battery indicator to work. When I contacted the folks at olarila for help they managed to get my battery indicator working but they use only full DSDT patching. never took the time out to create hot patches from it.
2) Honestly 16,3 has been more stable for me than 14,1 and 15,4 or 15,2 as well.
3) I used ResourceConverter.sh and edited Mac-E7203C0F68AA0004.plist (had to go through some old documentation fo edit the frequency vector).
4) They are useless. I have been trying to fix few Sleep/Wake issues causes by RTC wake ups so was testing something.

**Comment by benbender** [comment]
First, thanks for your response! I just realized that the discharge value seems to be sampled over time and doesn't seem to give a live-value. You left your system to idle quite some time to reach that point, right?

When I let my T480 idle for around 10 minutes, I also can reach about 4-5Watt.

Another factor seems to be hidpi, if I go down to 1280x720 instead of 1920x1080 (hidpi), the power usage seems to be dropping a bit.

**Comment by FlasHRender** [comment]
Lol. See ofc if you are working on laptop its going to consume more power. For me its reaches upto 6W on web surfing and higher depending on the task. 

From the thread I can see most of discussion are about idle power consumption. You can not compare power usage under load. 

I can do 10 other things to reduce power consumption lower than even that 4W but the screenshot is on a *useable* idle system(50% Brightness, Wifi and Bluetooth are on and connected, apps running in background, 1080p 60 fps resolution). I left the system to idle for 1 min and took a screenshot.

**Comment by benbender** [comment]
Sure, that's what I meant: mine needs ~5-10 minutes to reach that state and not after letting it idle for 1 or 2 minutes. That's what I was wondering about.

**Comment by FlasHRender** [comment]
> Sure, that's what I meant: mine needs ~5-10 minutes to reach that state and not after letting it idle for 1 or 2 minutes. That's what I was wondering about.

No no. I meant I let it sit for 1 min to cover the full 60 sec graph of the intel widget. But the power usage drops to 0.8W the second I just leave the laptop to idle and that's how it should be. Check for apps that aren't idling in your usage. There are no apps to find instant power draw other than Intel Widget so I'm assuming the sec Total pkg drops to 0.7W the total draw also drops to under 4W.

**Comment by benbender** [comment]
I was talking about the "discharging with"-value in coconut-battery in your screenshot. It originates from "AppleSmartBattery"-node in ioreg. That value reflects the discharge-rate of the battery and should therefore include things like display, usb, etc. Tuning of the CPU is only one part of the puzzle.

**Comment by tylernguyen** [comment]
Did some light testing of `MacbookAir8,1` includes matching framebuffer, usb remapping, cpufrienddataprovider
As far as I can tell, there was not much a different in power consumption or performance. This is after 2 days of normal usage and some light benchmarking with GeekBench and Cinebench.

So it seems that changing the SMBIOS will be a very minor if even effective part of dealing with power consumption. I will suggest looking into other aspects instead.

**Comment by benbender** [comment]
Yes and no. Did you regenerate CPUFriendProvider.kext with the changed Plattform-Id? Info.plist of the board is used as base to patch frequency-scaling etc. If you didn't, there wont be really that much visible difference. 

The real difference between plattforms are the values in the plists of the boards. To see what differs, have a direct look at: /System/Library/Extensions/IOPlatformPluginFamily.kext/Contents/PlugIns/X86PlatformPlugin.kext/Contents/Resources/.

But I'm +1 in the sense that this isn't the main point of tuning.

**Comment by tylernguyen** [comment]
> 
> 
> Yes and no. Did you regenerate CPUFriendProvider.kext with the changed Plattform-Id? Info.plist of the board is used as base to patch frequency-scaling etc. If you didn't, there wont be really that much visible difference.
> 
> The real difference between plattforms are the values in the plists of the boards. To see what differs, have a direct look at: /System/Library/Extensions/IOPlatformPluginFamily.kext/Contents/PlugIns/X86PlatformPlugin.kext/Contents/Resources/.
> 
> But I'm +1 in the sense that this isn't the main point of tuning.

@benbender as stated, I did. Also redid usbmapping since it the kext it generates matches the platform, and I also included matching framebuffer in testing so I am sure now that SMBIOS does not have any significant power consumption effect.

**Comment by benbender** [comment]
After some more research:
The Answer to the implications of the plattform-id is not clear because there are many (small) moving parts to the equation and the implications changed over time. 

- The powermanagement of the gpu(s) **IS** depending on the choice of the plattform. The values of our iGPU-PM can be seen under "AGPMController" and "AppleGraphicsDevicePolicy" in ioreg. The latter seems to be more important with configurations with a working dGPU.

- In the past there **WAS** a significant involvement of the plattform-id in cpu-scaling and thermal management, but that can be completely altered nowadays by CPUFriend/one-key-cpufriend (which injects essentially only parameters to X86PlatformPlugin and can be viewed in ioreg). **BUT** most of the tools to generate CPUFriendProvider.kext or similar are depending on the plattform-choice. The direct powerconsumption should be less effected, but it should over time because of different scaling-profiles/optimizations between different plattforms. 

Additionally there are many small parts and apps which are using the plattform-id to manage assumptions about the hw. F.e. naming of sensors in iStat.

So, in conclusion, I'll heading back to a macbookpro14,2-configuration with a CPUFriendProvider.kext generated on macbookpro15,2 and I will try to patch my system to be as closely as possible to match the macbookpro14,2-config.

**Comment by zombillano** [comment]
> After some more research:
> The Answer to the implications of the plattform-id is not clear because there are many (small) moving parts to the equation and the implications changed over time and with the usage of HWP on modern Intel CPUs.
> 
>     * The powermanagement of the gpu(s) **IS** depending on the choice of the plattform. The values of our GPU-PM can be seen under "AGPMController" and "AppleGraphicsDevicePolicy" in ioreg. The latter seems to be more important with configurations with a working dGPU.
> 
>     * In the past there **WAS** a significant involvement of the plattform-id in cpu-scaling and thermal management, but that can be completely altered nowadays by CPUFriend/one-key-cpufriend (which injects essentially only parameters to X86PlatformPlugin and can be viewed in ioreg). **BUT** most of the tools to generate CPUFriendProvider.kext or similar are depending on the plattform-choice. The direct powerconsumption should be less effected, but it is over time because of different scaling-profiles/optimizations between different plattforms.
> 
>     * There could be more, but I couldn't find other references by scanning original macbook's ioreg.
> 
> 
> Additionally there are many small parts and apps which are using the plattform-id to manage assumptions about the hw. F.e. naming of sensors in iStat.
> 
> So, in conclusion, I'll heading back to a macbookpro14,2-configuration with a CPUFriendProvider.kext generated on macbookpro15,2 and I will try to patch my system to be as closely as possible to match the macbookpro14,2-config.

T480 owner here. I would like to offer my help at testing (if you need it, of course). I have the one that comes with i7-8650u without dGPU and power management is one of the topics i've been looking every now and then (I've only got focussed on solving a problem that I'm having with my SSD). Hope I can help.

**Comment by velaar** [comment]
> CPUFriendProvider.kext generated on macbookpro15,2

@benbender can you please share the kext? I'd like to do some testing.

**Comment by benbender** [comment]
@velaar Just take one from the zip here: https://github.com/tylernguyen/x1c6-hackintosh/issues/28#issuecomment-649120727

**Comment by tylernguyen** [comment]
An update on the issue, I confirmed that there's a BIOS mod working on the x1c6, see Issue #68 
I assume that the unlocked BIOS will give us more options to control power usage/consumption on the machine. I just finished the mod today so I am still exploring the advance menu. But give it a shot and look around.

**Comment by zombillano** [comment]
Does this open the door to CFG Unlock? I read somewhere that it is needed by macOS for proper power managment

**Comment by benbender** [comment]
@tylernguyen I don't think that it will improve much. Aspm is only enabled for NVME + WIFI on genuin macs. Confirmed in ioreg available here: https://github.com/khronokernel/DarwinDumped/blob/master/MacBookPro/MacBookPro14%2C1/

@zombillano Basically yes. But there are far more options around that topic and it is unclear (at least to me) what os x really expects.

**Comment by tylernguyen** [comment]
Confirmed that the modded BIOS's undervolt menu/options are working. This is particular great because macOS doesn't really have any nice tools for doing so right now.

**Comment by zombillano** [comment]
Wasn't voltageshift working for you? Either way, I believe that BIOS undervolting is a better aproach. I already ordered a CH431A programmer and will try to mod my BIOS as soon as it arrives.

**Comment by tylernguyen** [comment]
> 
> 
> Wasn't voltageshift working for you? Either way, I believe that BIOS undervolting is a better aproach. I already ordered a CH431A programmer and will try to mod my BIOS as soon as it arrives.

I tried it and got it to a semi-working state but it has a lot of issues, particularly with the kext and sleep. And yeah, BIOS undervolt is just much better anyway.

**Comment by zombillano** [comment]
I saw the medium article that you posted. Then I found who wrote it, the good n4ru :D I will link you to this post at reddit about unlocking the TDP on a T480: https://www.reddit.com/r/thinkpad/comments/g8fk51/t480_consuming_60w_85w_total_unlimited_tdp/

45W TDP is insane, but it would be interesting to see if we can get sustained 3.9GHz even on high loads with something like 30W TDP or so. T480 benefits of being able to mod with dual heatpipe on the models without dGPU, which I believe gives the oportunity to the T480 to really squeeze out all the performance from the i7-8650u. I happen to have that exact model, so I will certainly give it a try soon. How are the thermals on the X1C6?

**Comment by tylernguyen** [comment]
> 
> 
> I saw the medium article that you posted. Then I found who wrote it, the good n4ru :D I will link you to this post at reddit about unlocking the TDP on a T480: https://www.reddit.com/r/thinkpad/comments/g8fk51/t480_consuming_60w_85w_total_unlimited_tdp/
> 
> 45W TDP is insane, but it would be interesting to see if we can get sustained 3.9GHz even on high loads with something like 30W TDP or so. T480 benefits of being able to mod with dual heatpipe on the models without dGPU, which I believe gives the oportunity to the T480 to really squeeze out all the performance from the i7-8650u. I happen to have that exact model, so I will certainly give it a try soon. How are the thermals on the X1C6?

I think thermals are fairly limited on the X1C6. The machine is pretty stuck with cooling it's got. There some remedies to be done:
* If you have the `Furukawa/Nidec` fan, you should replace it with a `Delta` one: https://www.reddit.com/r/thinkpad/comments/c7zpah/x1_carbon_6th_gen_horrible_cooling_fan_design/
* Repaste the machine with a better compound or liquid metal.

I've done the two things above and honestly, the machine still gets pretty warm sometimes. But honestly. right now I'm trying to reduce power consumption on battery. Increasing power consumption on power for better performance is another issue for another day.

**Comment by tylernguyen** [comment]
With some advanced BIOS tweaking, I was able to get idle power consumption pretty slow, pretty much identical to Linux/Windows.

![Screenshot_2020-07-20 17 10 07_PNIOGL](https://user-images.githubusercontent.com/3349081/87991546-07aa2c00-caac-11ea-88b6-ee25529c082d.png)

All that's left is to experiment and play around with real world usage on  battery.

**Comment by parndt** [comment]
@tylernguyen that's great news; can you please share your current BIOS settings?

**Comment by tylernguyen** [comment]
@parndt Sure, it is a working draft so it's subjected to experimentation and change:

* SpeedStep mode for Battery `Battery optimized`
* Adaptive Thermal Management for Battery `Balanced`

* Overclocked Features enabled to allow for undervolting. My personal stable profile are as followed:
  * processor= -95
  * gt slice and unslice = -80
  * uncore = -80

* CFG Lock `disabled`
* FCLK Frequency for Early power on `400MHz`
* Disable External Graphs support all around and especially in the TB3 section if you do not need it.
* Additionally, if you do not use TB3 hotplug at all on macOS, you should disable `Force Power` related options in TB3. (My example above had these enabled, so you should see improves with them disabled)
* TPD Boot Mode `Down`, this is at the cost of performance. However, there is nothing preventing you from creating a custom power limit profile. In fact, I hope that someone who has more time can experiment with a custom "balanced" power limit profile and share it as I simply do not have the time myself to do so. In addition, I rarely use my x1c6 for sustained heavy computing tasks so I have no need for the `Up` option.

* Under PCH-IO/USB:
You can manually disable any USB pin you do not use. I myself disabled the ones for the SD card reader as I never use it. This feature is also pretty cool since you can hard disable webcam/mic for privacy if you're extra cautious or simply never use the laptop's webcam anyway.

* Set DeepSX to `Enabled in S3-S4-S5/Battery`

A lot of these settings may be unnecessary but I am still experiment as this is only my second day with the unlocked BIOS. I will continue to edit this post as I experiment with more and different options.

**Comment by rjocoleman** [comment]
Have you played with DPTF? I've have some good results just turning it off with no other changes. It's the topic of a long thread on the Lenovo forums regarding the "throttling issue" faced by these machines.

**Comment by tylernguyen** [comment]
> 
> 
> Have you played with DPTF? I've have some good results just turning it off with no other changes. It's the topic of a long thread on the Lenovo forums regarding the "throttling issue" faced by these machines.

@rjocoleman

I have not. Could you link that thread?
I'm more focused on power consumption right now. Experimenting with thermal throttling requires stress testing with Prime95 which takes time so I'm inclined to leave it for later.

**Comment by rjocoleman** [comment]
https://forums.lenovo.com/t5/Other-Linux-Discussions/X1C6-T480s-low-cTDP-and-trip-temperature-in-Linux/m-p/4028489 (around page 20 something they start into the DPTF)
https://www.reddit.com/r/thinkpad/comments/870u0a/t480s_linux_throttling_bug/
https://github.com/erpalma/throttled

I note that many of the MSRs/trip points mentioned in those various threads are configurable in these menus.

**Comment by tylernguyen** [comment]
> 
> 
> https://forums.lenovo.com/t5/Other-Linux-Discussions/X1C6-T480s-low-cTDP-and-trip-temperature-in-Linux/m-p/4028489 (around page 20 something they start into the DPTF)
> https://www.reddit.com/r/thinkpad/comments/870u0a/t480s_linux_throttling_bug/
> https://github.com/erpalma/throttled
> 
> I note that many of the MSRs/trip points mentioned in those various threads are configurable in these menus.

Hm... I will try playing around with those variables. @rjocoleman Do you have any experience with tweaking power limit profiles? I myself do not.

**Comment by tylernguyen** [comment]
https://adamgradzki.com/2020/04/03/thinkpad-gpu-undervolt/

According to this article, apparently you can undervolt the gpu on the x1c6 as much as  -400 mV before seeing diminishing returns? That seems absurd but I guess I'll test it anyway.

**Comment by tylernguyen** [comment]
Also, Intel Ark https://ark.intel.com/content/www/us/en/ark/products/124968/intel-core-i7-8650u-processor-8m-cache-up-to-4-20-ghz.html says that the T junction for my i7-8650u is 100 C. So I am certain the default throttling/tripping points can be pushed. I'm just not sure what's the safe point for us to configure is at.

**Comment by benbender** [comment]
@tylernguyen wow :)

a) Would you be so kind to provide an AppleIntelInfo.kext-dump from Hackintool (SIP must be disabled) to see in which state the processor reached that state? (Hackintool -> Utilities -> Get Appleinfo, the most left icon on the bottom - as noted, SIP must be deactivated to load AppleIntelInfo.kext)

From my understanding atm its mostly about cpu-cores C-States reaching down to C-10 while the machine is running. It would be very helpful to pin the config needed for this.

b) I'm experimenting with S0-DeepIdle. With the new options, it can be activated and I'm trying to find the right combination to make it useable. ATM S0-Deepidle works, but screws S3 sleep in terms of usability... More Infos here: https://github.com/daliansky/OC-little/tree/master/01-%E5%85%B3%E4%BA%8EAOAC (You can confirm the principle ability to use S0-DeepIdle in the FACP-Table of ACPI via MacIASL)

**Comment by tylernguyen** [comment]
@benbender 
Here's the AppleIntelInfo dump
[appleintelinfo.txt](https://github.com/tylernguyen/x1c6-hackintosh/files/4956927/appleintelinfo.txt)

Yup, the power state is very stable. The system boosts normally upon tasks and otherwise idle below 1. And as far as I can tell, sleep is working normally.

* I have not been able to get S0-Deepdle working. It seems there's a bug in the BIOS mod that will not allow me to save settings for S0-Deepidle, it would just reset on boot.

**Comment by benbender** [comment]
The Bios a bit flaky in general when it comes to saving different values. F.e. I'm having I2C2-I2C5-devices activated since yesterday and I'm not able to deactivate them.

From your AppleIntelInfo:
`````
MSR_CONFIG_TDP_NOMINAL...........(0x648) : 0x13
MSR_CONFIG_TDP_LEVEL1............(0x649) : 0x80050
MSR_CONFIG_TDP_LEVEL2............(0x64a) : 0x1500C8
MSR_CONFIG_TDP_CONTROL...........(0x64b) : 0x1
MSR_TURBO_ACTIVATION_RATIO.......(0x64c) : 0x0
MSR_PKGC3_IRTL...................(0x60a) : 0x884E
MSR_PKGC6_IRTL...................(0x60b) : 0x8876
MSR_PKGC7_IRTL...................(0x60c) : 0x8894
MSR_PKG_C2_RESIDENCY.............(0x60d) : 0x28AF432DF
MSR_PKG_C3_RESIDENCY.............(0x3f8) : 0x4265AB643
MSR_PKG_C2_RESIDENCY.............(0x60d) : 0x28AF432DF
MSR_PKG_C3_RESIDENCY.............(0x3f8) : 0x4265AB643
MSR_PKG_C6_RESIDENCY.............(0x3f9) : 0x12223FD85
MSR_PKG_C7_RESIDENCY.............(0x3fa) : 0x13AF739
MSR_PKG_C8_RESIDENCY.............(0x630) : 0xB94B4DEF
MSR_PKG_C9_RESIDENCY.............(0x631) : 0x0
MSR_PKG_C10_RESIDENCY............(0x632) : 0x0
MSR_PKG_C8_LATENCY...............(0x633) : 0xB94B4DEF
MSR_PKG_C9_LATENCY...............(0x634) : 0x0
MSR_PKG_C10_LATENCY..............(0x635) : 0x0
````

vs. mine:

````
MSR_CONFIG_TDP_NOMINAL...........(0x648) : 0x13
MSR_CONFIG_TDP_LEVEL1............(0x649) : 0x80050
MSR_CONFIG_TDP_LEVEL2............(0x64a) : 0x1500C8
MSR_CONFIG_TDP_CONTROL...........(0x64b) : 0x80000000
MSR_TURBO_ACTIVATION_RATIO.......(0x64c) : 0x80000012
MSR_PKGC3_IRTL...................(0x60a) : 0x884E
MSR_PKGC6_IRTL...................(0x60b) : 0x8876
MSR_PKGC7_IRTL...................(0x60c) : 0x8894
MSR_PKG_C2_RESIDENCY.............(0x60d) : 0x577C77DA715
MSR_PKG_C3_RESIDENCY.............(0x3f8) : 0x2A45585FA05
MSR_PKG_C2_RESIDENCY.............(0x60d) : 0x577C77DA715
MSR_PKG_C3_RESIDENCY.............(0x3f8) : 0x2A45585FA05
MSR_PKG_C6_RESIDENCY.............(0x3f9) : 0x0
MSR_PKG_C7_RESIDENCY.............(0x3fa) : 0x0
MSR_PKG_C8_RESIDENCY.............(0x630) : 0x0
MSR_PKG_C9_RESIDENCY.............(0x631) : 0x0
MSR_PKG_C10_RESIDENCY............(0x632) : 0x0
MSR_PKG_C8_LATENCY...............(0x633) : 0x0
MSR_PKG_C9_LATENCY...............(0x634) : 0x0
MSR_PKG_C10_LATENCY..............(0x635) : 0x0
````

I don't know how to achive that config reliably. That's the key to get the CPU to power down in S0 while running. Do you use the CPUFriendProvider-config from this repo atm?

**Comment by tylernguyen** [comment]
@benbender Yup, I'm using the CPUFriendDataProvider from my repo currently.
I'm not too comfortable with sharing my BIOS dump publicly since it contains my machine's serial information. All the settings I've made are all documented above. But I'll send it to you personally if you contact me via Signal or email. And as far as I can tell, the biggest change in power is disabling CFG Lock and setting TDP config to `Down`

**Comment by benbender** [comment]
The problem is that we don't have a good baseline because "Loading defaults" doesn't reset everything in the bios. F.e. my I2C-Devices are there after a reset... ^^ And I'm using your settings afaik. There is surely something more to be done. I'll contact you at contact at tylerspaper dot com.

**Comment by benbender** [comment]
I'm now also able to push my pkg-usage below 1W. From my observations

> * CFG Lock `disabled`
needs to be manually checked with verifyMSR-efi-tool.

> * FCLK Frequency for Early power on `400MHz`
Doesn't have an impact for me.

> * Disable External Graphics support all around and especially in the TB3 section if you do not need it.
I can't switch my primary gfx from SG (switchable GFX) to iGPU, but I can manually configure PCIE and disable Port 1, which is my Nvidia-Card. Huge impage, it seems.

> * Additionally, if you do not use TB3 hotplug at all on macOS, you should disable `Force Power` related options in TB3. (My example above had these enabled, so you should see improvements with them disabled). This way, you would need to plug all TB3 devices at boot when needed, but would save power since the TB3 controller does not need to be powered on all the times.
+1

> * TPD Boot Mode `Down`, this is at the cost of performance. However, there is nothing preventing you from creating a custom power limit profile. In fact, I hope that someone who has more time can experiment with a custom "balanced" power limit profile and share it as I simply do not have the time myself to do so. In addition, I rarely use my x1c6 for sustained heavy computing tasks so I have no need for the `Up` option.
Doesn't have an impact for me.

> * Under PCH-IO/USB:
>   You can manually disable any USB pin you do not use. I myself disabled the ones for the SD card reader as I never use it. This feature is also pretty cool since you can hard disable webcam/mic for privacy if you're extra cautious or simply never use the laptop's webcam anyway.
+1

> * Set DeepSX to `Enabled in S3-S4-S5/Battery`
Doesn't have an impact for me.

DeepSX isn't used on OSX afaik.

**Comment by tylernguyen** [comment]
> 
> 
> I'm now also able to push my pkg-usage below 1W. From my observations
> 
> > * CFG Lock `disabled`
> 
> needs to be manually checked with verifyMSR-efi-tool.
> 

Hm.. so I guess there might be a bug where settings are not applied? Have you tried the reset method @velaar posted?

**Comment by benbender** [comment]
> Hm.. so I guess there might be a bug where settings are not applied? Have you tried the reset method @velaar posted?

Yes, and its hard to tell, if it worked. I can't get reliable results from my testing in general. Sometimes settings are not applied and then, out of nowhere, they are. For CFG Lock it was set, but VerifyMSR reported "MSR Locked" until it doesnt after some testing - no clue why... At this point I can only advise to try to double check...

I'm on vacation next week and will try a clean patch afterward to try to sort the stability issues.

**Comment by tylernguyen** [comment]
@benbender 

I found that the biggest power different is enabling Thunderbolt BIOS assist and enabling support in pre boot.

Please test those. When I have disabled, I'm experience about 2W while with them enabled, I'm able to get sub 1W.

**Comment by 1Revenger1** [comment]
If your wanting to look at iGPU power states, interrupts a second from various sources, and CPU usage, then you could try using `sudo powermetrics --show-process-energy` and see if you can find anything making a lot of interrupts. 

I've been looking through here for PM tips for my x1 - what is this thunderbolt support mode?

**Comment by tylernguyen** [comment]
@1Revenger1 
Posted on Lenovo's forum:
* In BIOS Assist Mode, memory resource arraignment is done by BIOS. 
* In Native mode, memory resource arrangement is done by OS.
And just as @veelar said:
In this generation of machine, Thunderbolt controller causes a C-storn on our laptops ramping the CPU package to 2.2-2.5W
If you disable it (or put into BIOS-assisted mode), the C-storm ends and the CPU properly clocks down to where it is supposed to be allowing us to reach a way better power state.

https://forums.lenovo.com/t5/Linux-Discussion/FYI-Linux-May-Not-Support-Thunderbolt-Native-Mode/td-p/4057604?page=1

**Comment by velaar** [comment]
@tylernguyen my speculation is that in BIOS assist mode the power management is also done by BIOS. (including USB-C/TB link power management). I can not confirm it but with proper Linux/Windows driver it is possible to achieve similar levels of power consumption. 

I wonder if the recent BIOS hacks will allow us to use OSX PM for TBT / USB-C

**Comment by FlasHRender** [comment]
@velaar there is a detailed guide on getting macOS PM for TBT. Here : https://osy.gitbook.io/hac-mini-guide/details/thunderbolt-3-fix-part-3

Either ways I will be trying the different BIOS options as suggested by Tyler. I use setup_var to change my bios settings as I don't have a fully modded bios chip. But I have been enable to setup all bios options via setup_var method. FYI I'm using Zenbook S UX391UA very similar specs to the laptop in discussion.

**Comment by velaar** [comment]
@FlasHRender 
I'm well aware of this fix, however I have a concern about using the updated TBT FW with Windows. As I depend on eGPU for my home office setup where both Windows and MacOS X are needed. 

I will say though that at least hot plug with BIOS assist  → on; should be a big improvement for my use case. 
But we digress, this is a discussion for TBT thread.

**Comment by tylernguyen** [comment]
A year and half after this issue was opened by @velaar, I feel that I am now ready to close this issue as I consider it solved. Here are my closing comments:
* The biggest culprit of idle power draw is the Thunderbolt 3 controller. With `Thunderbolt BIOS Assist` **disabled**, idle CPU power consumption is at `~2W` as opposed to `~0.8W` with the option **enabled**, a huge difference in real world usage. This is true for both Windows and macOS.
* However, this is especially troublesome for macOS/x1c6-hackintosh because:
  * As of right now, my TB3 hotplug patch will not work with `Thunderbolt BIOS Assist` **enabled**. This means that those who want TB3 hotplug capabilities will have to sacrifice battery time, and a decent amount of it too.
  * Furthermore, current hackintosh software fixies for Thunderbolt 3 hotplug requires that the the TB3 controller power be forced on at boot. This means that the controller is always powered, even when you do not need it.
  * Hence, you will have to choose between TB3 hotplug and power consumption for the time being on hackintosh. I personally have chosen battery time. I use TB3 quite often but am content with shutting down the machine each to plug in TB3 devices (a fast boot drive helps).

With that said, however, there are other things that can be done to improve CPU power consumption:
* First things first, you should disable all peripherals you do not need at the BIOS level, things such as: **always on USB**, fingerprint reader, wwan slot, camera, and microphone.
* Next, there are a few things that should be done by all to optimize CPU power draw on macOS, all of which requires tampering with the machine physically:
  * Repaste the machine with an aftermarket thermal paste to improve temps.
  * Mod the BIOS to unlock Intel Advance Menu.
  * Disable CFG Lock to allow for native power management.
  * Undervolt the machine using the Overclocking Menu within the modded BIOS.
  * Configure TDP to `Down` (comes at the cost of sustained performance).

Personally, with all those adjustments made and with `Thunderbolt BIOS Assist` **enabled**, my machine idles at `~0.8W`, which is identical to Windows and is good enough for a day's worth on medium brightness (by which my 1440p screen is the real power hog).

All of this can be viewed in more details starting the next commit and will be continually updated in my `docs/`, specifically `1_README-HARDWAREandBIOS.md` and `5_README-other.md`.

My next hope is that we can tackle the big boss, Thunderbolt 3 hotplug. Right now, it works perfectly but as previously mentioned, comes at a cost of the CPU idle power draw being `~2W`. I want to be able to have Thunderbolt 3 hotplug and low idle power draw too, damn it! See Issue #24 for further discussions on this topic.

Finally, thank you to everyone here, your interest is why I keep this project actively maintained.

**Comment by Porco-Rosso** [comment]
Hi @tylernguyen just a quick question, it might seem obvious. If we enable Thunderbolt BIOS Assist, this would also mean that to use TB3 on windows it would need to be plugged in at boot to work? 
I hardly use the GPU on MacOS but would definitely need it on windows for some light gaming. 

Lastly it would be great to have some documentation on how to patch the BIOS. I haven't dived into it yet, but would appreciate a starting point, and recommended settings.

**Comment by tylernguyen** [comment]
@Porco-Rosso 

Windows works fine with Thunderbolt BIOS assist enabled, only macOS is affected by this. Also, BIOS documentation will be included on the next repo commit. I'm waiting for a few final things before pushing it. Though I think it will be up in a less than one week.

**Comment by abbasabidii** [comment]
> If your wanting to look at iGPU power states, interrupts a second from various sources, and CPU usage, then you could try using `sudo powermetrics --show-process-energy` and see if you can find anything making a lot of interrupts.
> 
> I've been looking through here for PM tips for my x1 - what is this thunderbolt bios assist?

I know this is a different Hackintosh forum but I am having the same issues with Comet Lake i7-10610u CPU (Dell Latitude 7410) and these are the interrupts, if anyone can help what is causing them or how to solve them would be helpful. I know HDEF is related to sound I2C is related to Trackpad? and CNVW is wifi, XHC is USB but idk what is TMR. iStat shows 9.5 Watt CPU consumption on idle. 

```shell
****  Interrupt distribution ****

CPU 0:
	Vector 0x51(I2C1/DSB1/HDEF): 1888.65 interrupts/sec
	Vector 0x70(XHC): 41.90 interrupts/sec
	Vector 0x72(PXSX): 217.28 interrupts/sec
	Vector 0x77(CNVW): 21.95 interrupts/sec
	Vector 0x7a(IGPU): 7.78 interrupts/sec
	Vector 0xdd(TMR): 201.51 interrupts/sec
	Vector 0xde(IPI): 40.10 interrupts/sec
CPU 1:
	Vector 0xde(IPI): 6.58 interrupts/sec
CPU 2:
	Vector 0xdd(TMR): 22.15 interrupts/sec
	Vector 0xde(IPI): 34.92 interrupts/sec
CPU 3:
	Vector 0xde(IPI): 11.57 interrupts/sec
CPU 4:
	Vector 0xdd(TMR): 7.58 interrupts/sec
	Vector 0xde(IPI): 33.32 interrupts/sec
CPU 5:
	Vector 0xde(IPI): 8.98 interrupts/sec
CPU 6:
	Vector 0xdd(TMR): 5.59 interrupts/sec
	Vector 0xde(IPI): 24.14 interrupts/sec
CPU 7:
	Vector 0xdd(TMR): 1.40 interrupts/sec
	Vector 0xde(IPI): 4.59 interrupts/sec
```

**Comment by Mifaxav** [comment]
Hey, im on an X1 yoga gen 3, very close hardware to your machine. How can I decrease the power usage? On idle, power consumption is 8 watts total but the CPU is at 0.2 watts. PKG is at 1.5 - 1.8, dram at 0.4
